### ISSUE TITLE
Figure out how to return errors from callback interfaces

### ISSUE THREAD

**Issue by jhugman** [mozilla/uniffi-rs#351]
We need to be able to errors in foreign language callback interfaces, and pass them back to Rust.

The developer experience of this should be symmetrical with how errors are reported/passed for `interface`s.

┆Issue is synchronized with this [Jira Task](https://mozilla-hub.atlassian.net/browse/UNIFFI-38)
┆Issue Number: UNIFFI-38

**Comment by rfk** [comment]
Some brief straw thoughts on how this might work...

Currently we define the `ForeignCallback` interface via:

```
unsafe extern "C" fn(handle: u64, method: u32, args: RustBuffer) -> RustBuffer;
```

To allow the callback to report errors, this could grow an extra ` err: &mut ExternError` argument, symmetrically to how each FFI function exposed by the Rust code has an additional `err` argument:

```
unsafe extern "C" fn(handle: u64, method: u32, args: RustBuffer, err: &mut ExternError) -> RustBuffer;
```

When calling the foreign-language callback, the Rust code would need to create an `ExternError` struct and pass a mutable reference to it. Once the callback completes, the Rust code would need to check for a non-zero `code` on the error struct and propagate the error if so. To begin with we could have it simply panic on error, but a future iteration might allow us to convert the numeric code into a declared UDL error type and propagate it up into the user-written Rust code.

On the foreign-language side, we'd need to have a `try-catch` when dispatching the callback, and have it set the `code` and `message` fields on the error struct in the case of error. To begin with we could have it simply return a fixed error code (I think we use `-1` for unpexected errors?) but a future iteration might allow us to case-match on a declared UDL error type and have that convert into an Error instance in the Rust code.

**Comment by jplatte** [comment]
> On the foreign-language side, we'd need to have a `try-catch` when dispatching the callback,

Does that mean that currently, when a callback throws and doesn't catch the thrown exception it is up to Rust's `catch_unwind` that's part of every thread's first stack frames to deal with that exception? AFAIK that would mean the process is aborted.

It would be nice if UniFFI converted foreign exceptions to Rust panics instead.

**Comment by bendk** [comment]
Right now they're converted to panics, which are handled in different ways:
  - If you set `panic=abort` then the process gets aborted.  This is how FF Desktop works, which is why I'm going to try to prioritize this one so we can safely use callbacks there.
  - If not and the callback call originated from a normal UniFFI call then our `catch_unwind` code will catch it and convert it into an internal exception on the foreign side.
  - If not and the callback call originated from some other stack, then anything could happen, but probably an abort.

I think any change of behavior here would require a change on the Rust side as well.  The methods on the trait would need to return a `Result<>`, if not I think the only reasonable choice is to panic.  I think that we should define an enum something like this:

 ```
enum UniFFICallbackError {
   // Handles cases where the foreign code throws
   ForeignException(String),
   // Handles internal errors like failure to unpack the rust buffer.
   InternalError(String),
```

**Comment by mhammond** [comment]
> I think any change of behavior here would require a change on the Rust side as well. The methods on the trait would need to return a `Result<>`

Could we insist on that, with a deprecation warning like we did for threadsafe?

**Comment by bendk** [comment]
Implemented in #1344
